### DISCUSSION TITLE
Testers wanted - Ezra Bible App for Android mobile devices

### DISCUSSION THREAD

**Discussion by tobias-klein** [ezra-bible-app/ezra-bible-app#692] [Development]
I am currently working on [Ezra Bible App for Android mobile devices](https://github.com/ezra-bible-app/ezra-bible-app/issues/669).

It is the same app that is already available for tablets and bigger screens (like Chromebooks or Android TV). If you would like to participate in a test I would need your Google email address. As testers you will then get access to the test version via Google Play Store.

I would be glad if you could help me with this! You can contact me at [contact@ezrabibleapp.net](mailto:contact@ezrabibleapp.net).

**Comment by alerque** [comment]
I can't promise a lot of testing as I'm pretty swamped, but I can give it a spin on a couple devices: `caleb@alerque.com` should be registered as a Google user, but if it balks at that `alerque@gmail.com`.

**Reply by tobias-klein** [reply]
Thanks! I'll wait for a few more responses and then upload a current snapshot for testing.

**Comment by debritto** [comment]
Hi @tobias-klein !
I'm in turbulent moment at my work, but I'm willing to participate. My e-mail is: chriseana@gmail.com

**Comment by tobias-klein** [comment]
@alerque @debritto Any feedback? :)

**Comment by alerque** [comment]
Last I checked G Play wasn't letting me install in on either of my phones, only a tablet. I'll check again when I get a chance, the last time was a couple days ago (well after getting the beta invite link). I assumed it wasn't published yet..

**Reply by alerque** [reply]
Nope still a no go on my phones.

**Reply by tobias-klein** [reply]
Could you try with this link:
https://play.google.com/apps/testing/net.ezrabibleapp.cordova

If it still does not work, please delete the cache of the Google Play Store app.

**Reply by alerque** [reply]
I was eventually able to get started (that link some fussing with Google Play). I did get as far as one Bible working—but on initial setup I dried to download several and the others got lost. Since then attempts to add more modules have sometimes worked so I suppose there could have been a transient network issue.

The UI is clearly not optimized for mobile but I've been able to navigate it just fine and didn't bump into anything that outright didn't work (was cut off, not pressable, etc.).

**Reply by tobias-klein** [reply]
Thank you so much for testing and sharing feedback! :) Speaking about mobile screen optimization - is there anything specific that you would suggest to enhance?

**Reply by alerque** [reply]
The biggest things I noticed were very small targets to hit with fingers (comment/label flags next to verses for example, toolbar buttons that are small relative to the whitespace in the toolbars around them, lots of nested boxes (message in a frame in a dialog in a frame...) that ate up usable real estate, etc.). Nothing show stopper mind you, but a mobile friendly style overhaul would definitely be in order in usage picks up.

**Reply by tobias-klein** [reply]
Ok, I hear you. Thanks! I appreciate the feedback.

Regarding the "comment/label flags next to verses". This option should be turned off by default on mobile. I need to check why that was not the case.

The toolbar buttons and the spacing could already be enhanced now with some CSS changes.
The nested boxes ... I'll check whether some quick fix can be done. Otherwise certainly in one of the next releases. A pre-condition may be to first introduce a more modern JS/CSS framework (currently it's still jquery ui, which is pretty outdated, yet working), see #250.

**Comment by reyespinosa1996** [comment]
I will be more than happy to participate. You already have my email. Greetings.

**Reply by tobias-klein** [reply]
I just sent you an email with some info regarding the test version!

**Comment by Bukunmikuti** [comment]
Great work here! Count me in. 
My email: bukunmikuti@gmail.com

**Reply by tobias-klein** [reply]
I sent you an email!

**Comment by tobias-klein** [comment]
@reyespinosa1996 @Bukunmikuti @debritto
Any additional feedback on this?
I am planning to release next weekend. Obviously it is not going to be perfect from a style perspective, but if there are bugs like cut off areas, unpressable buttons, etc. that should be fixed.

**Comment by Bukunmikuti** [comment]
I used Ezra for one of my study sessions last week and I noted some issues during usage but I kept delaying this report, your mention was the reminder I needed.

So I'm just going to paste my findings here. I'll be reviewing as much as a regular user than as a developer so the proposed solutions here might appear subjective.
- The app onboarding went well. I would have preferred to have a default Bible version packaged with the app rather than having to set up one after installation. It makes getting started a lot easier. 
- I was able to search through the available languages but the search input wasn't available for Bible versions. I had to scroll to select my preferred version.
- The main screen is divided into two sections: the Bible reader and tag/dictionary section. It would be nice if the bible reader takes the whole screen with an option to toggle the other sections when needed. I only found out much later that I can actually make it hide. It should be hidden by default (?). Better still, the tag/dictionary section can be taken off the home screen to its own dedicated screen.
- The buttons are not entirely intuitive. I had to test each to see where it leads to. I would suggest a minimum number of buttons on the home screen and a menu that reveals other sections of the app with appropriate labels.
- There is no way to reference a scripture specifically rather than skipping through the chapters. It makes parallel cross referencing difficult. It would be nice if clicking on the chapter would bring options to select other chapters
- When a verse is highlighted, another click should remove the highlight rather than having to tap outside of the Bible reader.
- The context popover that appears after clicking the search and font-resize button does not disappear when other part of the screen is tapped (the expected UX) except the Bible reading section

### Bugs
- I started to install some translations and I cancelled one when it has already started. The progress bar showed completed but the application refused to continue with other the installation. I had to restart the app
- I made some notes on the app during my study. I check the next day and I discovered that they were all gone. I thought I had lost them all until I decided to select one of the verses and edit again. So apparently, saved notes are only visible after editing. Relaunching the app 'hides' the note until you click the edit button again. I would really prefer the note to be on a separate screen where I can edit and save as I like, even different notes for the same verse. There should only be an indication that I have a note saved to a particular verse as I read. Long notes under the verse can make the reader appear busy. 
---
- Night mode is awesome! The tags and compare function also works great. ✨✨
- The app runs smoothly, no lag. Great work so far. 

### NEW UPDATE
The app is refusing to launch properly. It is stuck at "initializing user interface". I have a fairly stable connection so I don't know what the problem could be. I tried clearing app data but to no avail. 

**Device:**
Redmi Note 11
128/4 GB
Android 11
@tobias-klein

**Reply by tobias-klein** [reply]
Thank you @Bukunmikuti for the detailed feedback. This is very helpful!

> The app onboarding went well. I would have preferred to have a default Bible version packaged with the app rather than having to set up one after installation. It makes getting started a lot easier.

Ok. I'll think about pre-packacking the KJV maybe?

> I was able to search through the available languages but the search input wasn't available for Bible versions. I had to scroll to select my preferred version.

Good point. I created issue #700 for that.

> The main screen is divided into two sections: the Bible reader and tag/dictionary section. It would be nice if the bible reader takes the whole screen with an option to toggle the other sections when needed. I only found out much later that I can actually make it hide. It should be hidden by default (?). Better still, the tag/dictionary section can be taken off the home screen to its own dedicated screen.

Yes, hiding the tool panel by default makes sense on mobile. I just fixed that in the code.

> The buttons are not entirely intuitive. I had to test each to see where it leads to. I would suggest a minimum number of buttons on the home screen and a menu that reveals other sections of the app with appropriate labels.

Thanks for this feedback. Yes, we can check how to improve this!

> There is no way to reference a scripture specifically rather than skipping through the chapters. It makes parallel cross referencing difficult. It would be nice if clicking on the chapter would bring options to select other chapters.

I just added issue #701 for this.

> When a verse is highlighted, another click should remove the highlight rather than having to tap outside of the Bible reader.

Yes, makes sense. This is currently not supported by the used JavaScript library. But I'll check anyway how to fix that. I added issue #702 for that.

> The context popover that appears after clicking the search and font-resize button does not disappear when other part of the screen is tapped (the expected UX) except the Bible reading section.

I'll look into this.

> I started to install some translations and I cancelled one when it has already started. The progress bar showed completed but the application refused to continue with other the installation. I had to restart the app.

I turned this into issue #703.

**Reply by Bukunmikuti** [reply]
> Ok. I'll think about pre-packacking the KJV maybe?

Yes, that would be nice!

**Comment by tobias-klein** [comment]
@Bukunmikuti 

> The app is refusing to launch properly. It is stuck at "initializing user interface". I have a fairly stable connection so I don't know what the problem could be. I tried clearing app data but to no avail.

Is this on the same device as earlier when it was working?
Some app data is stored under /sdcard/Documents/ezra. This may not be deleted based on your previous action. Could you rename the folder and then try again? Do you have a way to zip that folder and send it to me?

To find out what happened exactly I would have to supply you a debug version (then you can get me the debug logs). Would you mind contacting me via contact@ezrabibleapp.net?

**Reply by Bukunmikuti** [reply]
Renaming the folder fixed it.
I sent you an email

**Comment by tobias-klein** [comment]
Hi all!

@Bukunmikuti @alerque @debritto
Thanks for your feedback so far!

I have decided to postpone the next release to August, to have some more time to make it production ready. It seems like there are some show-stoppers specifically with Android 11 (which introduced the "scoped storage"). It's not that it doesn't run with Android 11, but the issues that have been observed definitely need to be adressed.

**Reply by alerque** [reply]
Fair enough.

Just for the record, my phone devices are on Android 12 (Lineage 19) and haven't noticed any issues in this genre. Are these issues `>= 11` or `< 12` or just `== 11`?

**Reply by tobias-klein** [reply]
I expect that these issues apply to anything `>= 11`. I have a switch in the code that actually differentiates based on the version number. The storage behavior / constraints changed with Android 11. It got quite complicated unfortunately with "scoped storage", see https://source.android.com/devices/storage/scoped?hl=en.
